src/deepsparse/v2/text_generation/autoregressive_preprocess_operator.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional
+
+import numpy
+
+from deepsparse.transformers.utils.helpers import create_causal_mask
+from deepsparse.v2.operators import Operator
+from deepsparse.v2.utils import Context, InferenceState, PipelineState
+
+
+__all__ = ["AutoRegressiveOperatorPreprocess"]
+
+
+class AutoRegressiveOperatorPreprocess(Operator):
+    def __init__(self, sequence_length: int, prompt_sequence_length: int):
+        """
+        Prepare the tokens for the single-token engine. This requires creating the
+        attention mask, positions, and causal mask. The output contains these three
+        arrays to be passed into the single-token engine.
+        """
+        self.sequence_length = sequence_length
+        self.prompt_sequence_length = prompt_sequence_length
+
+    def can_operate(self, inp: Any, context: Context, inference_state: InferenceState):
+        """
+        Can run this Operator if the number of tokens left to process is greater than
+        0 but less than the self.promt_sequence_length. Also, thie Operator can only
+        run after PrepareforSingleEngine as it requires the kv_cache to be updated.
+        """
+        tokens = inp.get("tokens")
+        kv_cache = inp.get("kv_cache")
+
+        found = False
+        for c in context.stages_executed:
```

Comment:
is there a way we can avoid requiring a specific operator to be run before, and instead do a check on the inputs?

Reply:
I know we're planning on updating the edges a bit so won't focus too much on this in this review

Reply:
updated to use just inputs. no longer checking context